File: epi/EvenOddListMerge.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

import java.util.List;
public class EvenOddListMerge {
  @EpiTest(testDataFile = "even_odd_list_merge.tsv")

  public static ListNode<Integer> evenOddMerge(ListNode<Integer> L) {
    /*
      Edge case: null return null
      Even first then odd
      //T O(N) S O(1)
     */
//    if(L == null) {
//      return null;
//    }
//    ListNode even = L, odd = L.next, oriOdd = odd;
//
//    while(true) {
//      if(odd == null || odd.next == null) {
//        even.next = oriOdd;
//        break;
//      }
//      even.next = odd.next;
//      even = even.next;
//      odd.next = even.next;
//      odd = odd.next;
//    }
//
//    return L;
    return epi.kt.EvenOddListMerge.INSTANCE.evenOddMerge(L);
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "EvenOddListMerge.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}

File: epi/kt/EvenOddListMerge.kt
package epi.kt

import epi.ListNode

object EvenOddListMerge {
    fun evenOddMerge(L: ListNode<Int>?) : ListNode<Int>? {
        if(L == null) {
            return null
        }

        var even = L
        var odd = even?.next
        val oH = odd

        while(true) {
            if(odd == null || odd.next == null) {
                even?.next = oH
                break
            }
            even?.next = odd.next
            even = even?.next
            odd.next = even?.next
            odd = odd.next
        }

        return L
    }

}